source/kernel/src/forth.rs

```diff
@@ -191,6 +299,147 @@ impl Params {
+/// allows for forth-bound builtin functions to retrieve the referred objects in a
+/// type safe way.
+pub struct BagOfHolding {
+    idx: i32,
+    inner: HeapFixedVec<(i32, BohValue)>,
+    kernel: &'static Kernel,
+}
+
+impl BagOfHolding {
+    /// Create a new BagOfHolding with a given max size
+    ///
+    /// The `kernel` parameter is used to allocate `HeapBox` elements to
+    /// store the type-erased elements residing in the BagOfHolding.
+    pub async fn new(kernel: &'static Kernel, max: usize) -> Self {
+        let inner = kernel.heap().allocate_fixed_vec(max).await;
+        BagOfHolding {
+            idx: 0,
+            inner,
+            kernel,
+        }
+    }
+
+    /// Generate a new, currently unused, Bag of Holding token.
+    ///
+    /// This token will never be zero, and a given bag of holding will never
+    /// contain two items with the same token.
+    fn next_idx(&mut self) -> i32 {
+        // todo we could do better lol
+        loop {
+            self.idx = self.idx.wrapping_add(1);
+            if self.idx == 0 {
+                continue;
+            }
+            if !self.inner.iter().any(|(idx, _)| *idx == self.idx) {
+                return self.idx;
+            }
+        }
+    }
+
+    /// Place an item into the bag of holding
+    ///
+    /// The item will be allocated into a `HeapBox`, and a non-zero i32 token will
+    /// be returned on success.
+    ///
+    /// Returns an error if the Bag of Holding is full.
+    ///
+    /// At the moment there is no way to "unregister" an item, it will exist until
+    /// the BagOfHolding is dropped.
+    pub async fn register<T>(&mut self, item: T) -> Result<i32, ()>
+    where
+        T: 'static,
+    {
+        if self.inner.is_full() {
+            return Err(());
         }
+        let leaked = self.kernel.heap().allocate(item).await.leak().cast::<()>();
+        let idx = self.next_idx();
+        let tid = TypeId::of::<T>();
+
+        // Should never fail - we checked whether we are full above already
+        self.inner
+            .push((
+                idx,
+                BohValue {
+                    tid,
+                    leaked,
+                    dropfn: dropfn::<T>,
+                },
+            ))
+            .map_err(drop)?;
+        Ok(idx)
+    }
+
+    /// Attempt to retrieve an item from the Bag of Holding
+    ///
+    /// This will only succeed if the same `T` is used as was used when calling
+    /// [`BagOfHolding::register()`], and if the token matches the one returned
+    /// by `register()`.
+    ///
+    /// If the token is unknown, or the `T` does not match, `None` will be returned
+    ///
+    /// At the moment, no `get_mut` functionality is provided. It *could* be, as the
+    /// Bag of Holding represents ownership of the contained items, however it would
+    /// not be possible to retrieve multiple mutable items at the same time. This
```

Comment:
is retrieving multiple bagged items mutably at the same time something we actually want to do? AFAICT, each bag of holding is uniquely owned by a forth VM instance, right?

Reply:
Yeah, it would be needed if you wanted to do something like:

```rust
let t2 = stack.pop()?;
let t1 = stack.pop()?;
let something = boh.get_mut(t1)?;
let other = boh.get_mut(t2)?;
something.do_thing(other)?;
```

```forth
t1 t2 something::do_thing
```